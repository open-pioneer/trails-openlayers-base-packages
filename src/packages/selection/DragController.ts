// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { Resource } from "@open-pioneer/core";
import OlMap from "ol/Map";
import { unByKey } from "ol/Observable";
import Overlay from "ol/Overlay";
import { mouseActionButton } from "ol/events/condition";
import Geometry from "ol/geom/Geometry";
import { DragBox, DragPan } from "ol/interaction";
import PointerInteraction from "ol/interaction/Pointer";

interface InteractionResource extends Resource {
    interaction: PointerInteraction;
}
/** Represents a tooltip rendered on the OpenLayers map. */
interface Tooltip extends Resource {
    overlay: Overlay;
    element: HTMLDivElement;
    setText(value: string): void;
}

const ACTIVE_CLASS = "selection-active";
const INACTIVE_CLASS = "selection-inactive";

export class DragController {
    private tooltip: Tooltip;
    private interactionResources: InteractionResource[] = [];
    private olMap: OlMap;
    private isActive: boolean = true;
    private tooltipMessage: string;
    private tooltipDisabledMessage: string;

    constructor(
        olMap: OlMap,
        tooltipMessage: string,
        tooltipDisabledMessage: string,
        onExtentSelected: (geometry: Geometry) => void
    ) {
        const viewPort = this.initViewport(olMap);
        this.interactionResources.push(
            this.createDragBox(olMap, onExtentSelected, viewPort, this.interactionResources)
        );
        this.interactionResources.push(this.createDrag(olMap, viewPort, this.interactionResources));

        this.tooltip = this.createHelpTooltip(olMap, tooltipMessage);
        this.olMap = olMap;
        this.tooltipMessage = tooltipMessage;
        this.tooltipDisabledMessage = tooltipDisabledMessage;
    }

    initViewport(olMap: OlMap) {
        const viewPort = olMap.getViewport();
        viewPort.classList.add(ACTIVE_CLASS);

        viewPort.oncontextmenu = (e) => {
            e.preventDefault();
            return false;
        };
        return viewPort;
    }

    /**
     * Method for destroying the controller when it is no longer needed
     */
    destroy() {
        this.tooltip.destroy();
        this.interactionResources.forEach((interaction) => {
            interaction.destroy();
        });
    }

    setActive(isActive: boolean) {
        if (this.isActive === isActive) return;
        const viewPort = this.olMap.getViewport();
        if (isActive) {
            this.interactionResources.forEach((interaction) =>
                this.olMap.addInteraction(interaction.interaction)
            );
            this.tooltip.setText(this.tooltipMessage);
            viewPort.classList.remove(INACTIVE_CLASS);
            viewPort.classList.add(ACTIVE_CLASS);
            this.isActive = true;
        } else {
            this.interactionResources.forEach((interaction) =>
                this.olMap.removeInteraction(interaction.interaction)
            );
            this.tooltip.setText(this.tooltipDisabledMessage);
            viewPort.classList.remove(ACTIVE_CLASS);
            viewPort.classList.add(INACTIVE_CLASS);
            this.isActive = false;
        }
    }

    /**
     * Method to create a simple extent-selection
     */
    private createDragBox(
        olMap: OlMap,
        onExtentSelected: (geometry: Geometry) => void,
        viewPort: HTMLElement,
        interactionResources: InteractionResource[]
    ): InteractionResource {
        const dragBox = new DragBox({
            className: "selection-drag-box",
            condition: mouseActionButton
        });

        olMap.addInteraction(dragBox);
        dragBox.on("boxend", function () {
            onExtentSelected(dragBox.getGeometry());
        });

        const interactionResource: InteractionResource = {
            interaction: dragBox,
            destroy() {
                olMap.removeInteraction(dragBox);
                interactionResources.splice(interactionResources.indexOf(this));
                dragBox.dispose();
                viewPort.classList.remove(ACTIVE_CLASS);
                viewPort.classList.remove(INACTIVE_CLASS);
                viewPort.oncontextmenu = null;
            }
        };
        return interactionResource;
    }

    /**
     * Method to activate pan with right-mouse-click
     */
    private createDrag(
        olMap: OlMap,
        viewPort: HTMLElement,
        interactionResources: InteractionResource[]
    ): InteractionResource {
        const condition = function (mapBrowserEvent: {
            originalEvent: MouseEvent;
            dragging: unknown;
        }) {
            const originalEvent = /** @type {MouseEvent} */ mapBrowserEvent.originalEvent;
            return originalEvent.button == 2;
        };
        const drag = new DragPan({
            condition: condition
        });

        olMap.addInteraction(drag);

        const interactionResource: InteractionResource = {
            interaction: drag,
            destroy() {
                olMap.removeInteraction(drag);
                interactionResources.splice(interactionResources.indexOf(this));
                drag.dispose();
                viewPort.classList.remove(ACTIVE_CLASS);
                viewPort.classList.remove(INACTIVE_CLASS);
                viewPort.oncontextmenu = null;
            }
        };

        return interactionResource;
    }

    /**
     * Method to generate a tooltip on the mouse cursor
     */
    private createHelpTooltip(olMap: OlMap, message: string): Tooltip {
        const element = document.createElement("div");
        element.className = "selection-tooltip printing-hide";
        element.role = "tooltip";

        const content = document.createElement("span");
        content.textContent = message;
        element.appendChild(content);

        const overlay = new Overlay({
            element: element,
            offset: [15, 0],
            positioning: "center-left"
        });

        const pointHandler = olMap.on("pointermove", (evt) => {
            overlay.setPosition(evt.coordinate);
        });

        olMap.addOverlay(overlay);
        return {
            overlay,
            element,
            destroy() {
                olMap.removeOverlay(overlay);
                overlay.dispose();
                unByKey(pointHandler);
            },
            setText(value) {
                content.textContent = value;
            }
        };
    }

    /**
     * Method for testing purposes only
     * @returns InteractionResource of class DragBox
     */
    getDragboxInteraction() {
        return this.interactionResources.find(
            (interactionResource) => interactionResource.interaction instanceof DragBox
        );
    }

    /**
     * Method for testing purposes only
     * @returns InteractionResource of class DragPan
     */
    getDragPanInteraction() {
        return this.interactionResources.find(
            (interactionResource) => interactionResource.interaction instanceof DragPan
        );
    }
}
