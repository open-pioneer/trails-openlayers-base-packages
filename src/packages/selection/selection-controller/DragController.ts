// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { Resource } from "@open-pioneer/core";
import OlMap from "ol/Map";
import { mouseActionButton } from "ol/events/condition";
import Geometry from "ol/geom/Geometry";
import { DragBox, DragPan } from "ol/interaction";
import PointerInteraction from "ol/interaction/Pointer";
import { activateViewportInteraction, createHelpTooltip, deactivateViewportInteraction, Tooltip } from "./helper";

interface InteractionResource extends Resource {
    interaction: PointerInteraction;
}

export class DragController {
    private tooltip: Tooltip;
    private interactionResources: InteractionResource[] = [];
    private olMap: OlMap;
    private isActive: boolean = true;
    private tooltipMessage: string;
    private tooltipDisabledMessage: string;

    constructor(
        olMap: OlMap,
        tooltipMessage: string,
        tooltipDisabledMessage: string,
        onExtentSelected: (geometry: Geometry) => void
    ) {
        const viewPort = this.initViewport(olMap);
        this.interactionResources.push(
            this.createDragBox(olMap, onExtentSelected, viewPort, this.interactionResources)
        );
        this.interactionResources.push(this.createDrag(olMap, viewPort, this.interactionResources));

        this.tooltip = createHelpTooltip(olMap, tooltipMessage);
        this.olMap = olMap;
        this.tooltipMessage = tooltipMessage;
        this.tooltipDisabledMessage = tooltipDisabledMessage;
    }

    initViewport(olMap: OlMap) {
        activateViewportInteraction(olMap);
        const viewPort = olMap.getViewport();

        viewPort.oncontextmenu = (e) => {
            e.preventDefault();
            return false;
        };
        return viewPort;
    }

    /**
     * Method for destroying the controller when it is no longer needed
     */
    destroy() {
        this.tooltip.destroy();
        this.interactionResources.forEach((interaction) => {
            interaction.destroy();
        });
        deactivateViewportInteraction(this.olMap, true);
    }

    setActive(isActive: boolean) {
        if (this.isActive === isActive) return;
        if (isActive) {
            this.interactionResources.forEach((interaction) =>
                this.olMap.addInteraction(interaction.interaction)
            );
            this.tooltip.setText(this.tooltipMessage);
            activateViewportInteraction(this.olMap);
            this.isActive = true;
        } else {
            this.interactionResources.forEach((interaction) =>
                this.olMap.removeInteraction(interaction.interaction)
            );
            this.tooltip.setText(this.tooltipDisabledMessage);
            deactivateViewportInteraction(this.olMap);
            this.isActive = false;
        }
    }

    /**
     * Method to create a simple extent-selection
     */
    private createDragBox(
        olMap: OlMap,
        onExtentSelected: (geometry: Geometry) => void,
        viewPort: HTMLElement,
        interactionResources: InteractionResource[]
    ): InteractionResource {
        const dragBox = new DragBox({
            className: "selection-drag-box",
            condition: mouseActionButton
        });

        olMap.addInteraction(dragBox);
        dragBox.on("boxend", function () {
            onExtentSelected(dragBox.getGeometry());
        });

        const interactionResource: InteractionResource = {
            interaction: dragBox,
            destroy() {
                olMap.removeInteraction(dragBox);
                interactionResources.splice(interactionResources.indexOf(this));
                dragBox.dispose();
                deactivateViewportInteraction(olMap);
                viewPort.oncontextmenu = null;
            }
        };
        return interactionResource;
    }

    /**
     * Method to activate pan with right-mouse-click
     */
    private createDrag(
        olMap: OlMap,
        viewPort: HTMLElement,
        interactionResources: InteractionResource[]
    ): InteractionResource {
        const condition = function (mapBrowserEvent: {
            originalEvent: MouseEvent;
            dragging: unknown;
        }) {
            const originalEvent = /** @type {MouseEvent} */ mapBrowserEvent.originalEvent;
            return originalEvent.button == 2;
        };
        const drag = new DragPan({
            condition: condition
        });

        olMap.addInteraction(drag);

        const interactionResource: InteractionResource = {
            interaction: drag,
            destroy() {
                olMap.removeInteraction(drag);
                interactionResources.splice(interactionResources.indexOf(this));
                drag.dispose();
                deactivateViewportInteraction(olMap);
                viewPort.oncontextmenu = null;
            }
        };

        return interactionResource;
    }


    /**
     * Method for testing purposes only
     * @returns InteractionResource of class DragBox
     */
    getDragboxInteraction() {
        return this.interactionResources.find(
            (interactionResource) => interactionResource.interaction instanceof DragBox
        );
    }

    /**
     * Method for testing purposes only
     * @returns InteractionResource of class DragPan
     */
    getDragPanInteraction() {
        return this.interactionResources.find(
            (interactionResource) => interactionResource.interaction instanceof DragPan
        );
    }
}
