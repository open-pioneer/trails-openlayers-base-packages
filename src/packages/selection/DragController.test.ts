// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { afterEach, expect, vi, it } from "vitest";
import { DragController } from "./DragController";
import OlMap from "ol/Map";

afterEach(() => {
    vi.restoreAllMocks();
});

it("expect tooltip to be successfully created after construction", async () => {
    const { olMap, tooltipTest } = createController();
    const activeTooltip = getTooltipElement(olMap, "selection-tooltip");
    expect(activeTooltip).toMatchSnapshot(tooltipTest);
});

it("expect extent handler to be called", async () => {
    const { controller, extentHandler } = createController();
    const dragBox = controller.getDragboxInteraction();
    if (!dragBox) throw new Error("myDragbox not found");
    dragBox.interaction.dispatchEvent("boxend");
    expect(extentHandler).toBeCalledTimes(1);
});

it("expect interactions, tooltip and cursor correspond to controller state", async () => {
    const { olMap, controller, tooltipTest, disabledTooltipText } = createController();
    const activeTooltip = getTooltipElement(olMap, "selection-tooltip");

    const dragBox = controller.getDragboxInteraction()?.interaction;
    const dragPan = controller.getDragPanInteraction()?.interaction;

    if (dragBox === undefined || dragPan === undefined)
        throw new Error("at least one interaction not found");

    expect(olMap.getInteractions().getArray()).contains(dragBox);
    expect(olMap.getInteractions().getArray()).contains(dragPan);
    expect(olMap.getViewport().classList.contains("selection-active")).toBeTruthy();
    expect(activeTooltip.textContent).toBe(tooltipTest);

    controller.setActive(false);

    expect(olMap.getInteractions().getArray()).not.contains(dragBox);
    expect(olMap.getInteractions().getArray()).not.contains(dragPan);
    expect(olMap.getViewport().classList.contains("selection-inactive")).toBeTruthy();
    expect(activeTooltip.textContent).toBe(disabledTooltipText);
});

function createController() {
    const olMap = new OlMap();
    const tooltipTest = "Tooltip wurde gesetzt";
    const disabledTooltipText = "Funktion ist deaktiviert";
    const extentHandler = vi.fn();
    const controller = new DragController(olMap, tooltipTest, disabledTooltipText, extentHandler);
    return { olMap, controller, tooltipTest, extentHandler, disabledTooltipText };
}

function getTooltipElement(olMap: OlMap, className: string): HTMLElement {
    const allOverlays = olMap.getOverlays().getArray();
    const tooltips = allOverlays.filter((ol) => ol.getElement()?.classList.contains(className));
    if (tooltips.length === 0) {
        throw Error("did not find any tooltips");
    }
    const element = tooltips[0]!.getElement();
    if (!element) {
        throw new Error("tooltip overlay did not have an element");
    }
    return element;
}
