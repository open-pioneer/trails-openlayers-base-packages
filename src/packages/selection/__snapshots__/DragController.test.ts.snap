// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expect tooltip to be successfully created after construction > Tooltip wurde gesetzt 1`] = `
<div
  class="selection-tooltip printing-hide"
>
  Tooltip wurde gesetzt
</div>
`;
